binr/r2r/r2r.c

```diff
@@ -0,0 +1,271 @@
+
+		r_th_lock_enter (state->lock);
+		r_pvector_push (&state->results, result);
+		switch (result->result) {
+		case R2R_TEST_RESULT_OK:
+			state->ok_count++;
+			break;
+		case R2R_TEST_RESULT_FAILED:
+			state->xx_count++;
+			break;
+		case R2R_TEST_RESULT_BROKEN:
+			state->br_count++;
+			break;
+		case R2R_TEST_RESULT_FIXED:
+			state->fx_count++;
+			break;
+		}
+		r_th_cond_signal (state->cond);
+	}
+	r_th_lock_leave (state->lock);
+	return R_TH_STOP;
+}
+
+static void print_diff(const char *actual, const char *expected) {
+	// TODO: do an actual diff
+	RList *lines = r_str_split_duplist (expected, "\n");
+	RListIter *it;
+	char *line;
+	r_list_foreach (lines, it, line) {
+		printf (Color_RED"-"Color_RESET" %s\n", line);
+	}
+	r_list_free (lines);
+	lines = r_str_split_duplist (actual, "\n");
+	r_list_foreach (lines, it, line) {
+		printf (Color_GREEN"+"Color_RESET" %s\n", line);
+	}
+	r_list_free (lines);
+}
+
+static void print_result_diff(R2RTestResultInfo *result) {
+	switch (result->test->type) {
+	case R2R_TEST_TYPE_CMD: {
+		const char *expect = result->test->cmd_test->expect.value;
+		if (!expect) {
+			expect = "";
+		}
+		if (strcmp (result->proc_out->out, expect) != 0) {
+			printf ("-- stdout\n");
+			print_diff (result->proc_out->out, expect);
+		}
+		expect = result->test->cmd_test->expect_err.value;
+		if (!expect) {
+			break;
+		}
+		if (strcmp (result->proc_out->err, expect) != 0) {
+			printf ("-- stderr\n");
+			print_diff (result->proc_out->err, expect);
+		}
+		break;
+	}
+	case R2R_TEST_TYPE_ASM:
+		// TODO
+		break;
+	case R2R_TEST_TYPE_JSON:
+		break;
+	}
+}
+
+static void print_state(R2RState *state, ut64 prev_completed) {
+	printf ("\r\x1b[2K");
+
+	// Detailed test result (with diff if necessary)
+	ut64 completed = (ut64)r_pvector_len (&state->results);
+	ut64 i;
+	for (i = prev_completed; i < completed; i++) {
+		R2RTestResultInfo *result = r_pvector_at (&state->results, (size_t)i);
+		if (!state->verbose && (result->result == R2R_TEST_RESULT_OK || result->result == R2R_TEST_RESULT_FIXED)) {
+			continue;
+		}
+		char *name = r2r_test_name (result->test);
+		if (!name) {
+			continue;
+		}
+		switch (result->result) {
+		case R2R_TEST_RESULT_OK:
+			printf (Color_GREEN"[OK]"Color_RESET);
+			break;
+		case R2R_TEST_RESULT_FAILED:
+			printf (Color_RED"[XX]"Color_RESET);
+			break;
+		case R2R_TEST_RESULT_BROKEN:
+			printf (Color_BLUE"[BR]"Color_RESET);
+			break;
+		case R2R_TEST_RESULT_FIXED:
+			printf (Color_CYAN"[FX]"Color_RESET);
+			break;
+		}
+		printf (" %s "Color_YELLOW"%s"Color_RESET"\n", result->test->path, name);
+		if (result->result == R2R_TEST_RESULT_FAILED || (state->verbose && result->result == R2R_TEST_RESULT_BROKEN)) {
+			print_result_diff (result);
+		}
+		free (name);
+	}
+
+	// [x/x] OK  42 BR  0 ...
+	int w = printf ("[%"PFMT64u"/%"PFMT64u"]", completed, (ut64)r_pvector_len (&state->db->tests));
+	while (w >= 0 && w < 20) {
```

Comment:
use r_str_pad

Reply:
No, `r_str_pad()` uses a static variable and must die.